File: desktop/electron/services/exportFormat.ts
import type { ChatMessage, Conversation, Persona } from '../../../shared/types';
import { parseImageMessage } from '../../../shared/types';

export type ExportFormat = 'markdown' | 'json';

function roleLabel(message: ChatMessage, personaById: Map<string, Persona>): string {
  if (message.role === 'user') return 'User';
  const persona = message.personaId ? personaById.get(message.personaId) : undefined;
  return persona ? persona.name : 'Assistant';
}

export function formatConversationMarkdown(
  conversation: Conversation,
  messages: ChatMessage[],
  personas: Persona[],
): string {
  const personaById = new Map(personas.map((persona) => [persona.id, persona]));
  const lines = [`# ${conversation.title}`, ''];

  for (const message of messages) {
    const image = message.role === 'assistant' ? parseImageMessage(message.content) : null;
    lines.push(`**${roleLabel(message, personaById)}:**`, '');
    if (image) {
      lines.push(`_Generated image: ${image.prompt}_ (${image.model}, ${image.size})`, '');
    } else {
      lines.push(message.content, '');
    }
  }

  return lines.join('\n');
}

export function formatConversationJson(
  conversation: Conversation,
  messages: ChatMessage[],
): string {
  return JSON.stringify({ conversation, messages }, null, 2);
}

export function slugifyTitle(title: string): string {
  const slug = title
    .trim()
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-+|-+$/g, '')
    .slice(0, 60);
  return slug || 'conversation';
}

File: desktop-java/src/main/java/com/multiagent/desktop/service/ExportFormat.java
package com.multiagent.desktop.service;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.multiagent.desktop.model.ChatMessage;
import com.multiagent.desktop.model.Conversation;
import com.multiagent.desktop.model.MessageRole;
import com.multiagent.desktop.model.Persona;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Formats a conversation for export, mirroring desktop/electron/services/exportFormat.ts.
 * Image messages ([[MA_IMAGE]] payloads) aren't special-cased here the way the TS version
 * does, since image generation (Phase 3) hasn't been ported yet - there's nothing yet that
 * could produce one in the Java app's own conversations.
 */
public final class ExportFormat {
    private ExportFormat() {
    }

    public static String toMarkdown(Conversation conversation, List<ChatMessage> messages, List<Persona> personas) {
        Map<String, Persona> personaById = new LinkedHashMap<>();
        for (Persona persona : personas) {
            personaById.put(persona.getId(), persona);
        }

        StringBuilder sb = new StringBuilder();
        sb.append("# ").append(conversation.getTitle()).append("\n\n");
        for (ChatMessage message : messages) {
            sb.append("**").append(roleLabel(message, personaById)).append(":**\n\n");
            sb.append(message.getContent()).append("\n\n");
        }
        return sb.toString();
    }

    private static String roleLabel(ChatMessage message, Map<String, Persona> personaById) {
        if (message.getRole() == MessageRole.USER) {
            return "User";
        }
        Persona persona = message.getPersonaId() != null ? personaById.get(message.getPersonaId()) : null;
        return persona != null ? persona.getName() : "Assistant";
    }

    public static String toJson(Conversation conversation, List<ChatMessage> messages) {
        try {
            ObjectMapper mapper = new ObjectMapper();
            Map<String, Object> payload = new LinkedHashMap<>();
            payload.put("conversation", conversation);
            payload.put("messages", messages);
            return mapper.writerWithDefaultPrettyPrinter().writeValueAsString(payload);
        } catch (Exception e) {
            throw new IllegalStateException("Failed to serialize conversation to JSON", e);
        }
    }

    public static String slugifyTitle(String title) {
        String slug = title == null ? "" : title.trim().toLowerCase()
                .replaceAll("[^a-z0-9]+", "-")
                .replaceAll("^-+|-+$", "");
        if (slug.length() > 60) {
            slug = slug.substring(0, 60);
        }
        return slug.isEmpty() ? "conversation" : slug;
    }
}
